src/fasta/faidx.cpp

```diff
@@ -0,0 +1,199 @@
+/**
+ * Copyright 2019 EMBL - European Bioinformatics Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <fstream>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "fasta/faidx.hpp"
+
+struct ContigIndex {
+    std::string name;     // Name of this contig
+    uint64_t seq_length;  // Total length of this contig, in bases
+    uint64_t seq_offset;  // Offset in the FASTA file of this contig's first base
+    uint32_t line_length; // The number of bases on each line
+    uint32_t line_bytes;  // The number of bytes in each line, including the newline
+};
+
+class FastaIndexerState;
+
+class FastaIndexer {
```

Comment:
Well, it may appear to be like a missile, in my eyes it is a swiss army knife. For parsing a fasta file, yes, either way is clear and simple. I just felt a bit more comfortable with separating out each states.
I have put in another change in aa50e83 to address the (pseudo-)recursive calls. I didn't like them either but I don't know why I did not figure out the iterative way. Maybe it was because the fact that most (if not all) fasta files do not contain a lot of contigs made me relax with the call stacks. Anyway, with the new change this issue should have been resolved.